packages/liveblocks-core/src/lib/AsyncCache.ts

```diff
@@ -0,0 +1,345 @@
+   */
+  get(key: string): Promise<AsyncResolvedState<TData, TError>>;
+
+  /**
+   * Returns the current state of the key synchronously.
+   *
+   * @param key The key to get the state of.
+   */
+  getState(key: string): AsyncState<TData, TError> | undefined;
+
+  /**
+   * Marks a key as invalid, which means that the next
+   * {@link AsyncCache.get} call will re-invoke the function.
+   *
+   * @param key The key to invalidate.
+   * @param options.clearData Whether to clear the cached data.
+   * @param options.setOptimisticData Set data immediately but rollback if there's an error.
+   */
+  invalidate(key: string, options?: InvalidateOptions<TData>): void;
+
+  /**
+   * Calls {@link AsyncCache.invalidate} and then {@link AsyncCache.get}.
+   *
+   * @param key The key to revalidate.
+   * @param options.clearData Whether to clear the cached data.
+   * @param options.setOptimisticData Set data immediately but rollback if there's an error.
+   */
+  revalidate(
+    key: string,
+    options?: InvalidateOptions<TData>
+  ): Promise<AsyncResolvedState<TData, TError>>;
+
+  /**
+   * Subscribes to a key's changes.
+   *
+   * @param key The key to subscribe to.
+   * @param callback The function invoked on every change.
+   */
+  subscribe(
+    key: string,
+    callback: Callback<AsyncState<TData, TError>>
+  ): UnsubscribeCallback;
+
+  /**
+   * Returns whether a key already exists in the cache.
+   *
+   * @param key The key to look for.
+   */
+  has(key: string): boolean;
+
+  /**
+   * Clears all keys.
+   */
+  clear(): void;
+};
+
+const noop = () => {};
+
+export function isDifferentState(a: AsyncState, b: AsyncState): boolean {
+  // This might not be true, `data` and `error` would have to be
+  // deeply compared to know that. But in our use-case, `data` and
+  // `error` can't change without being set to `undefined` first or
+  // `isLoading` also changing in between or at the same time.
+  return (
+    a.isLoading !== b.isLoading ||
+    (a.data === undefined) !== (b.data === undefined) ||
+    (a.error === undefined) !== (b.error === undefined)
+  );
+}
+
+function createCacheItem<TData = any, TError = any>(
+  key: string,
+  asyncFunction: AsyncFunction<TData>,
+  { deduplicationInterval }: AsyncCacheItemOptions
+): AsyncCacheItem<TData, TError> {
+  const context: AsyncCacheItemContext<TData, TError> = {
+    isLoading: false,
+    isInvalid: true,
+    previousState: { isLoading: false },
+  };
+  const eventSource = makeEventSource<AsyncState<TData, TError>>();
+
+  function notify() {
+    const state = getState();
+
+    // We only notify subscribers if the cache has changed.
+    if (isDifferentState(context.previousState, state)) {
+      context.previousState = state;
+      eventSource.notify(state);
+    }
+  }
+
+  function invoke() {
+    context.lastInvokedAt = Date.now();
+    context.error = undefined;
+    context.isLoading = true;
+    context.isInvalid = true;
+    context.promise = asyncFunction(key);
```

Comment:
Rather than "mutating properties" on the context one by one, the benefit of using the 3-legged ADT and having all of its members use _readonly_ properties, is that you cannot do that. The only way to change from one state to another (e.g. from "loading" to "resolved") is to create a new object whenever you change to a new state.

Writing your code that way massively reduces the opportunity for bugs to lurk in there, and makes reasoning about the possible states easier, because there are literally only 3 states to consider.

Right now, for example, I see this `invoke()` will change 5 properties on `context`, but… I don't know (without look at all the other code), if there are perhaps other properties that must also be reset but perhaps have been forgotten, or if there are properties that deliberately remain the same.

Writing this as a state change using the ADT method, you'd write it as either:

```ts
// Option A: these are literally the only 5 fields on there! Easy to see. All the rest is wiped!
context = {
  lastInvokedAt: Date.now(),
  error: undefined,
  isLoading: true,
  isInvalid: true,
  promise: asyncFunction(key),
}
```

or

```ts
// Option B: these are literally the only 5 fields on there that are changed. All the rest remains the same. Also easy to see.
context = {
  lastInvokedAt: Date.now(),
  error: undefined,
  isLoading: true,
  isInvalid: true,
  promise: asyncFunction(key),
  ...context,
}
```

Another benefit is that if you (or someone else) later must introduce another property on the `context` object, then TypeScript will help you remember to consider/update the impact of that new property on this method. Is it relevant here in `invoke`, or not?

Reply:
I separated the state from the context and moved to always create a new state when modifying it, but the context is more of a collection of variables so I'm not sure if it would make sense to separate it into a discriminated union of possibilities.